### PR TITLE
Update toolchain to Go 1.23.1

### PR DESCRIPTION
**Pull request by peterebden**
Forgot in #3248 that this would be necessary too.

### CODE REVIEW

**Review by toastwaffle — APPROVED** [review]

**Comment by peterebden** [comment]
Looks like something's up with the coverage code on the alt build. Think that will need a proper fix.

File: src/test/coverage_test.go [hunk 1]
```diff
@@ -12,10 +12,11 @@ import (
 var target = &core.BuildTarget{Label: core.BuildLabel{PackageName: "src/test", Name: "coverage_test"}}
 
 const (
-	pythonCoverageFile    = "src/test/test_data/python-coverage.xml"
-	goCoverageFile        = "src/test/test_data/go_coverage.txt"
-	goCoverageFile2       = "src/test/test_data/go_coverage_2.txt"
-	goCoverageFile3       = "src/test/test_data/go_coverage_3.txt"
+	pythonCoverageFile = "src/test/test_data/python-coverage.xml"
+	// TODO(peterebden): Remove the 'x' on these once we update go-rules again
```
Comment (peterebden) [inline src/test/coverage_test.go:16]: I think there's an issue here where the coverage variables we create have the prefix `goCover` and so do these; they later get somehow detected as coverage vars. I guess something in 1.23 must have changed to find them.

The easiest / more correct fix is probably to make go-rules use a less usual coverage variable name. I'm changing these for now so I don't have to block on another round of updates there.

File: plugins/BUILD [hunk 2]
```diff
@@ -1,7 +1,8 @@
 plugin_repo(
     name = "go",
     plugin = "go-rules",
-    revision = "v1.20.1",
+    revision = "02f5910e44bb8274b5bdb80ec80188a9d180f59f",
```
Comment (chrisnovakovic) [inline plugins/BUILD:4]: Need to revert this to the official go-rules before merging.

**Review by goddenrich — APPROVED** [review]

File: plugins/BUILD (hunk 2 above)
Comment (peterebden) [reply plugins/BUILD]: Yup thanks, have done - wanted to get a build on here against it to check that really did fix it.

**Review by chrisnovakovic — APPROVED** [review]

**Pull request merged**